pkg/buildqueue/buildqueue.go

```diff
@@ -0,0 +1,580 @@
+// apply to exclusive builds.
+func numWorkerThreads() int {
+	ncpu := runtime.NumCPU()
+
+	if ncpu < minWorkerThreads {
+		return minWorkerThreads
+	}
+	return ncpu
+}
+
+// Buildable is an object that is buildable
+type Buildable interface {
+	// GetUUID must return a unique UUID of the object
+	GetUUID() string
+}
+
+// Builder is an object that can build itself. A builder must also be Buildable
+type Builder interface {
+	Buildable
+
+	// BuildQueued is called every time a builder is enqueued
+	BuildQueued()
+
+	// BuildsDequeued is called when a scheduled builder has been dequeued
+	// for building or has been cancelled. The nbuilds accounts for the
+	// number of builders folded together while the builder was queued.
+	BuildsDequeued(nbuilds int, cancelled bool)
+
+	// Build must build the Buildable. When a builder was cancelled,
+	// Build() will not be called. When a builder is folded into an
+	// existing builder, Build() is only called on the first builder.
+	Build() error
+}
+
+// BuildNotification is a channel to receive the status of a build. The channel
+// will receive true if the build was successful. The channel is buffered and
+// is guaranteed to only be written to exactly once and is then immediately
+// closed. It is up to the caller of Enqueue() to decide whether to read on the
+// channel or not.
+type BuildNotification <-chan bool
+
+func newBuildNotification() chan bool {
+	return make(chan bool, 1)
+}
+
+// build is what is returned by Enqueue() to provide information about the
+// build
+type build struct {
+	builder Builder
+
+	// uuid is the UUID of the Buildable
+	uuid string
+
+	// notificationChannels is a list of a channels that should be notified
+	// about the completion of the build
+	notificationChannels []chan bool
+}
+
+// createAndAddNotificationChannel creates and adds a build notification
+// channel. This function may only be called while the build is not yet in
+// progress.
+func (b *build) createAndAddNotificationChannel() BuildNotification {
+	notify := newBuildNotification()
+	b.notificationChannels = append(b.notificationChannels, notify)
+	return notify
+}
+
+// reportStatus reports the status of a build to all notification channels.
+// This function may only be called if the build is the currentBuild. It may
+// never be called while the build is still assigned to nextBuild.
+func (b *build) reportStatus(q *BuildQueue, success bool) {
+	for _, notify := range b.notificationChannels {
+		notify <- success
+		close(notify)
+	}
+}
+
+// cancel cancels a scheduled and not yet running build. q.mutex must be held.
+func (b *build) cancelScheduled(q *BuildQueue) int {
+	numBuilds := b.numFolded()
+
+	metrics.BuildQueueEntries.With(map[string]string{
+		metrics.LabelBuildQueueName: q.name,
+		metrics.LabelBuildState:     metrics.BuildStateWaiting,
+	}).Sub(float64(numBuilds))
+
+	b.reportStatus(q, false)
+
+	return numBuilds
+}
+
+// numFolded returns the number of folded builds
```

Comment:
This isn't blocking, but a lot of the function documentation in this PR is self-referential and thus doesn't provide much value. Given that this is not an exported function, it's not a big deal, though.